pkg/suggestion/v1beta1/hyperopt/base_service.py

```diff
@@ -62,14 +64,76 @@ def create_hyperopt_domain(self):
-                )
-            elif param.type == DOUBLE:
-                hyperopt_search_space[param.name] = hyperopt.hp.uniform(
-                    param.name, float(param.min), float(param.max)
-                )
+            if param.type in [INTEGER, DOUBLE]:
+                if param.distribution == api_pb2.UNIFORM or param.distribution is None:
+                    if param.step:
+                        hyperopt_search_space[param.name] = hyperopt.hp.quniform(
+                            param.name,
+                            float(param.min),
+                            float(param.max),
+                            float(param.step),
+                        )
+                    else:
+                        if param.type == INTEGER:
+                            hyperopt_search_space[param.name] = hyperopt.hp.uniformint(
+                                param.name, float(param.min), float(param.max)
+                            )
+                        else:
+                            hyperopt_search_space[param.name] = hyperopt.hp.uniform(
+                                param.name, float(param.min), float(param.max)
+                            )
+                elif param.distribution == api_pb2.LOG_UNIFORM:
+                    if param.step:
+                        hyperopt_search_space[param.name] = hyperopt.hp.qloguniform(
+                            param.name,
+                            math.log(float(param.min)),
+                            math.log(float(param.max)),
+                            float(param.step),
+                        )
+                    else:
+                        hyperopt_search_space[param.name] = hyperopt.hp.loguniform(
+                            param.name,
+                            math.log(float(param.min)),
+                            math.log(float(param.max)),
+                        )
+                elif param.distribution == api_pb2.NORMAL:
+                    mu = (float(param.min) + float(param.max)) / 2
+                    # We consider the normal distribution based on the range of ±3 sigma.
+                    sigma = (float(param.max) - float(param.min)) / 6
+
+                    if param.step:
+                        hyperopt_search_space[param.name] = hyperopt.hp.qnormal(
+                            param.name,
+                            mu,
+                            sigma,
+                            float(param.step),
+                        )
+                    else:
+                        hyperopt_search_space[param.name] = hyperopt.hp.normal(
+                            param.name,
+                            mu,
+                            sigma,
+                        )
+                elif param.distribution == api_pb2.LOG_NORMAL:
+                    log_min = math.log(float(param.min))
+                    log_max = math.log(float(param.max))
```

Comment:
Shouldn't we use the fixed value when the min and max are scalers the same as Nevergrad, right?

Reply:
Yeah we can, but that was an edge case when `min` and `max` are not defined in case of nevergrad.
```python
    elif isinstance(param, (p.Log, p.Scalar)):
        if (param.bounds[0][0] is None) or (param.bounds[1][0] is None):
            if isinstance(param, p.Scalar) and not param.integer:
                return hp.lognormal(label=param_name, mu=0, sigma=1)
``` 
For example,
```
    - name: batch_size
      parameterType: int
      feasibleSpace:
        min: "32"
        max: "64"
        distribution: "logNormal"
``` 
The above parameter will be sampled out from this graph:
<img width="681" alt="Screenshot 2024-09-22 at 8 48 49 PM" src="https://github.com/user-attachments/assets/cabea4e8-19ca-4e6e-8847-3089dcb26a93">
where `u=3.8123` and `sigma=0.3465` are calculated by putting `min=32` and `max=64` in our code. and `E(X)` represents the mean which is `48` in our case.

Reply:
That makes sense.
In that case, could you address the cases where min or max is not specified, as well as nevergrad?

https://github.com/facebookresearch/nevergrad/blob/a2006e50b068fe598e0f3d7dab9c9bcf6cf97e00/nevergrad/optimization/externalbo.py#L61-L64

Reply:
@shashank-iitbhu This is still pending.

Reply:
https://github.com/kubeflow/katib/blob/867c40a1b0669446c774cd6e770a5b7bbf1eb2f1/pkg/webhook/v1beta1/experiment/validator/validator.go#L287-L290

The webhook validator requires `feasibleSpace.max` or `feasibleSpace.min` to be specified.

Reply:
But when either min or max is empty, this validation does not reject the request, right?
So, shouldn't we implement the special case in the Suggestion Service?

Reply:
Yes, the validation webhook does not reject the request when either min or max is empty. But I created an example where:
```
    - name: batch_size
      parameterType: int
      feasibleSpace:
        min: "32"
        distribution: "logNormal"
``` 
For this, the experiment is being created but the suggestion service is not sampling out any value hence the trials are not running, though handled this case (when either min or max are not specified) in `pkg/suggestion/v1beta1/hyperopt/base_service.py`.
Do we need to check `experiment_defaults.go` file?
https://github.com/kubeflow/katib/blob/867c40a1b0669446c774cd6e770a5b7bbf1eb2f1/pkg/apis/controller/experiments/v1beta1/experiment_defaults.go